File: frontend/src/lib/ruleEngine.js
// JS port of backend/app/rule_engine.py — kept in lock-step with it so the
// app can classify locally (offline demo / backend down) exactly the same
// way the FastAPI backend does. If you change a threshold, change it in
// src/data/questionnaire.json (and the backend's copy) — not here.
import diet from '../data/diet_protocols.json'
import questionnaire from '../data/questionnaire.json'

const PHENOTYPE_IDS = Object.keys(questionnaire.phenotypes)

function get(values, field) {
  const v = values[field]
  if (v === undefined || v === null || v === '') return null
  return v
}

export function evaluateRule(rule, values) {
  const { op } = rule
  if (op === 'any') return rule.rules.some((r) => evaluateRule(r, values))
  if (op === 'all') return rule.rules.every((r) => evaluateRule(r, values))
  if (op === 'not') return !evaluateRule(rule.rule, values)

  const v = get(values, rule.field)

  if (op === 'eq') return v !== null && String(v).toLowerCase() === String(rule.value).toLowerCase()
  if (op === 'in') return v !== null && rule.values.map(String).includes(String(v))
  if (v === null) return false
  const n = Number(v)
  if (op === 'gt') return n > rule.value
  if (op === 'gte') return n >= rule.value
  if (op === 'lt') return n < rule.value
  if (op === 'lte') return n <= rule.value
  if (op === 'outside') {
    const [low, high] = rule.range
    return n < low || n > high
  }
  throw new Error(`Unknown rule op: ${op}`)
}

const FORMULAS = {
  bmi: (v) => v.weight_kg / Math.pow(v.height_cm / 100, 2),
  whr: (v) => v.waist_cm / v.hip_cm,
  tg_hdl_ratio: (v) => v.triglycerides / v.hdl,
}

// Live BMI helper for the Metabolic page (mentor feedback: show BMI as soon
// as height/weight are entered, not just after submitting). Not tied to the
// rule engine's internal `bmi` field name on purpose, so the UI can call it
// straight from raw form input.
export function liveBmi(heightCm, weightKg) {
  const h = Number(heightCm)
  const w = Number(weightKg)
  if (!h || !w) return null
  const bmi = w / Math.pow(h / 100, 2)
  if (!Number.isFinite(bmi)) return null
  let category = 'Normal'
  if (bmi < 18.5) category = 'Underweight'
  else if (bmi >= 25 && bmi < 30) category = 'Overweight'
  else if (bmi >= 30) category = 'Obese'
  return { bmi: Math.round(bmi * 10) / 10, category }
}

export function computeDerivedFields(values) {
  const enriched = { ...values }
  for (const phenotype of Object.values(questionnaire.phenotypes)) {
    for (const q of phenotype.questions) {
      const computed = q.computed || {}
      for (const key of Object.keys(computed)) {
        const fn = FORMULAS[key]
        if (!fn) continue
        try {
          const result = fn(enriched)
          if (Number.isFinite(result)) enriched[key] = result
        } catch {
          /* missing inputs -> leave uncomputed */
        }
      }
    }
  }
  return enriched
}

function severityLabel(positiveCount) {
  for (const band of questionnaire.severity_bands.bands) {
    if (positiveCount >= band.min && positiveCount <= band.max) return band.label
  }
  return 'Severe Driver / Primary Target'
}

export function scorePhenotypes(answers) {
  const values = computeDerivedFields(answers)
  const results = {}

  for (const [pid, phenotype] of Object.entries(questionnaire.phenotypes)) {
    let biochemicalPositive = 0
    const ticks = phenotype.questions.map((q) => {
      const positive = evaluateRule(q.rule, values)
      if (positive && q.category === 'Biochemical Evaluation') biochemicalPositive += 1
      return { id: q.id, label: q.label, category: q.category, reference: q.reference, positive }
    })
    const total = phenotype.questions.length
    const positiveCount = ticks.filter((t) => t.positive).length
    results[pid] = {
      label: phenotype.label,
      positive_count: positiveCount,
      total_questions: total,
      percentage: Math.round((positiveCount / total) * 1000) / 10,
      severity: severityLabel(positiveCount),
      biochemical_positive_count: biochemicalPositive,
      biomarker_support_flag: biochemicalPositive >= 3,
      ticks,
    }
  }

  const mito = questionnaire.mitochondrial_axis
  const mitoTicks = mito.questions.map((q) => ({
    id: q.id, label: q.label, reference: q.reference, positive: evaluateRule(q.rule, values),
  }))
  const mitoPositive = mitoTicks.filter((t) => t.positive).length
  results.mitochondrial = {
    label: mito.label,
    positive_count: mitoPositive,
    total_questions: mito.questions.length,
    percentage: Math.round((mitoPositive / mito.questions.length) * 1000) / 10,
    ticks: mitoTicks,
    flag: mitoPositive >= 2,
  }

  return { scores: results, values }
}

export function evaluateOverrides(values) {
  const triggered = []
  for (const override of questionnaire.overrides) {
    if (evaluateRule(override.rule, values)) {
      triggered.push({ id: override.id, phenotype: override.phenotype, label: override.label, reference: override.reference })
    }
  }
  return triggered
}

export function classify(scores, overridesTriggered) {
  const percentages = Object.fromEntries(PHENOTYPE_IDS.map((pid) => [pid, scores[pid].percentage]))
  const maxPct = Math.max(...Object.values(percentages))
  const tallyLeaders = PHENOTYPE_IDS.filter((pid) => percentages[pid] === maxPct)

  const overridePhenotypes = [...new Set(overridesTriggered.map((o) => o.phenotype))].sort()
  const biomarkerSupport = PHENOTYPE_IDS.filter((pid) => scores[pid].biomarker_support_flag)

  let classification, primary, involved, reason

  if (overridePhenotypes.length >= 2) {
    classification = 'Mixed/Combination Phenotype'
    primary = null
    involved = overridePhenotypes
    reason = `Multiple biomarker overrides triggered simultaneously (${overridesTriggered
      .map((o) => o.label)
      .join(', ')}) — Section 3 Key Biomarker Priority Overrules.`
  } else if (overridePhenotypes.length === 1) {
    const forced = overridePhenotypes[0]
    classification = 'Primary Phenotype (Biomarker Override)'
    primary = forced
    involved = [forced]
    const matching = overridesTriggered.find((o) => o.phenotype === forced)
    reason = `Biomarker override triggered: ${matching.label} — ${matching.reference}.`
    if (!tallyLeaders.includes(forced)) {
      reason += ` This overrules the tally leader(s) (${tallyLeaders.join(', ')}).`
    }
  } else if (maxPct === 0) {
    classification = 'Inconclusive / Minimal Findings'
    primary = null
    involved = []
    reason = 'No positive ticks recorded in any of the four phenotype categories.'
  } else if (tallyLeaders.length === 1) {
    classification = 'Primary Phenotype'
    primary = tallyLeaders[0]
    involved = [primary]
    reason = `Highest tick percentage (${maxPct}%) with no conflicting biomarker override.`
  } else {
    classification = 'Mixed/Combination Phenotype'
    primary = null
    involved = tallyLeaders
    reason = `Tied highest tick percentage (${maxPct}%) between ${tallyLeaders.join(', ')} — Section 3 Mixed/Combination rule.`
  }

  return {
    classification,
    primary_phenotype: primary,
    phenotypes_involved: involved,
    reason,
    tally_leaders: tallyLeaders,
    overrides_triggered: overridesTriggered,
    biomarker_supported_phenotypes: biomarkerSupport,
  }
}

function bodyTypeModifier(bodyType) {
  const modifiers = diet.body_type_modifiers
  if (!modifiers || !bodyType) return null
  return modifiers[bodyType] || null
}

function phenotypeBlock(phenotypeId, region, dietType, bodyType) {
  const general = diet.general_protocol[phenotypeId]
  const regional = diet.regional_suggestions[phenotypeId]
  const plan = diet.meal_plans[phenotypeId]
  const modifier = bodyTypeModifier(bodyType)
  return {
    phenotype: phenotypeId,
    focus: general.focus,
    diet_strategy: modifier ? [...general.diet_strategy, ...modifier.diet_strategy] : general.diet_strategy,
    exercise: modifier ? [...general.exercise, ...modifier.exercise] : general.exercise,
    priorities: general.priorities,
    body_type: bodyType || null,
    regional_suggestions: {
      region,
      primary_focus: regional.primary_focus,
      suggestions: regional[region] || {},
    },
    day_plan: {
      diet_type: dietType,
      meals: plan[dietType] || {},
      mid_meals_and_beverages: plan.mid_meals,
    },
  }
}

export function buildRecommendation(classification, scores, region, dietType, bodyType) {
  const involved = classification.phenotypes_involved
  const isMixed = classification.classification.startsWith('Mixed')
  const blocks = involved.map((pid) => phenotypeBlock(pid, region, dietType, bodyType))
  const mitoFlag = scores.mitochondrial?.flag
  return {
    is_mixed: isMixed,
    phenotype_blocks: blocks,
    mitochondrial_support: {
      recommended_priority: mitoFlag ? 'high' : 'foundational',
      note: mitoFlag
        ? 'Mitochondrial/hypoxia screen (Section 2, Table 5) flagged 2+ positive findings — treat this as a co-primary support track alongside the phenotype protocol above.'
        : 'General foundational support; add if energy, fatigue, or hypoxia symptoms are prominent.',
      ...phenotypeBlock('mitochondrial', region, dietType, bodyType),
    },
  }
}

export function checkRotterdamAndExclusions(answers) {
  if (answers.exclusion_other_disorders === 'yes') {
    return { status: 'excluded', reason: 'Excluded due to other suspected thyroid or pituitary disorders.' }
  }

  let rotterdamCount = 0
  if (answers.irregular_cycle === 'yes') rotterdamCount++
  if (answers.high_testosterone_symptoms === 'yes' || (answers.total_testosterone && Number(answers.total_testosterone) > 45)) rotterdamCount++
  if (answers.polycystic_ovaries_usg === 'yes') rotterdamCount++

  if (rotterdamCount < 2) {
    return { status: 'not_pcos', reason: 'Does not meet Rotterdam Criteria (requires 2 of 3: irregular cycles, high testosterone, polycystic ovaries).' }
  }

  return { status: 'pcos', reason: 'Meets Rotterdam Criteria for PCOS diagnosis.' }
}

export function runPipeline(answers, region, dietType) {
  const { scores, values } = scorePhenotypes(answers)
  
  const rotterdamResult = checkRotterdamAndExclusions(values)
  if (rotterdamResult.status !== 'pcos') {
    return {
      scores,
      classification: {
        classification: rotterdamResult.status === 'excluded' ? 'Excluded' : 'Inconclusive — Does not meet Rotterdam Criteria',
        primary_phenotype: null,
        phenotypes_involved: [],
        reason: rotterdamResult.reason,
        overrides_triggered: [],
        biomarker_supported_phenotypes: [],
      },
      recommendation: {
        is_mixed: false,
        phenotype_blocks: [],
        mitochondrial_support: null,
      },
      computed_values: values
    }
  }

  const overridesTriggered = evaluateOverrides(values)
  const classification = classify(scores, overridesTriggered)
  const recommendation = buildRecommendation(classification, scores, region, dietType, values.body_type)
  return { scores, classification, recommendation, computed_values: values }
}

export { questionnaire, diet as dietProtocols, PHENOTYPE_IDS }

File: backend/app/rule_engine.py
"""
Deterministic rule engine for PCOS phenotype classification (Plan A).

No AI/ML anywhere in this file. Every decision is a direct, traceable evaluation
of the JSON in app/data/questionnaire.json, which is itself a line-for-line
transcription of the uploaded questionnaire (see each field's "source").

Pipeline (mirrors the Work Plan's Steps 2-4 exactly):
    1. score_phenotypes()   -> Step 2: tally positive ticks per phenotype (/10)
    2. evaluate_overrides() -> Step 3: hard override rules on lab biomarkers
    3. classify()           -> Step 4: primary / mixed phenotype decision
"""
import json
import math
from pathlib import Path
from typing import Any, Dict, List, Optional

DATA_DIR = Path(__file__).parent / "data"

with open(DATA_DIR / "questionnaire.json") as f:
    QUESTIONNAIRE = json.load(f)

with open(DATA_DIR / "diet_protocols.json") as f:
    DIET_PROTOCOLS = json.load(f)

PHENOTYPE_IDS = list(QUESTIONNAIRE["phenotypes"].keys())  # adrenal, hormonal, inflammatory, metabolic


# ---------------------------------------------------------------------------
# Rule DSL interpreter -- the only "logic" allowed is these six operators, so
# every rule in the JSON is auditable by a non-programmer.
# ---------------------------------------------------------------------------
def _get(values: Dict[str, Any], field: str) -> Optional[float]:
    v = values.get(field)
    if v is None or v == "":
        return None
    return v


def evaluate_rule(rule: Dict[str, Any], values: Dict[str, Any]) -> bool:
    op = rule["op"]
    if op == "any":
        return any(evaluate_rule(r, values) for r in rule["rules"])
    if op == "all":
        return all(evaluate_rule(r, values) for r in rule["rules"])
    if op == "not":
        return not evaluate_rule(rule["rule"], values)

    field = rule.get("field")
    v = _get(values, field)

    if op == "eq":
        return v is not None and str(v).lower() == str(rule["value"]).lower()
    if op == "in":
        return v is not None and str(v) in [str(x) for x in rule["values"]]
    if v is None:
        # Numeric comparisons with no data entered yet cannot be positive.
        return False
    if op == "gt":
        return float(v) > rule["value"]
    if op == "gte":
        return float(v) >= rule["value"]
    if op == "lt":
        return float(v) < rule["value"]
    if op == "lte":
        return float(v) <= rule["value"]
    if op == "outside":
        low, high = rule["range"]
        return float(v) < low or float(v) > high
    raise ValueError(f"Unknown rule op: {op}")


# ---------------------------------------------------------------------------
# Safe arithmetic for the handful of computed fields (bmi, whr, tg_hdl_ratio)
# ---------------------------------------------------------------------------
import ast
import operator as _operator

_ALLOWED_BINOPS = {ast.Add: _operator.add, ast.Sub: _operator.sub,
                   ast.Mult: _operator.mul, ast.Div: _operator.truediv,
                   ast.Pow: _operator.pow}


def _safe_eval(expr: str, names: Dict[str, float]) -> Optional[float]:
    tree = ast.parse(expr, mode="eval")

    def _ev(node):
        if isinstance(node, ast.Expression):
            return _ev(node.body)
        if isinstance(node, ast.BinOp) and type(node.op) in _ALLOWED_BINOPS:
            return _ALLOWED_BINOPS[type(node.op)](_ev(node.left), _ev(node.right))
        if isinstance(node, ast.Num):  # py<3.8 fallback
            return node.n
        if isinstance(node, ast.Constant):
            return node.value
        if isinstance(node, ast.Name):
            val = names.get(node.id)
            if val is None:
                raise KeyError(node.id)
            return float(val)
        if isinstance(node, ast.UnaryOp) and isinstance(node.op, ast.USub):
            return -_ev(node.operand)
        raise ValueError(f"Disallowed expression: {ast.dump(node)}")

    try:
        return _ev(tree)
    except (KeyError, ZeroDivisionError, TypeError):
        return None


def compute_derived_fields(values: Dict[str, Any]) -> Dict[str, Any]:
    """Adds bmi/whr/tg_hdl_ratio etc. to a copy of `values` wherever inputs exist."""
    enriched = dict(values)
    for phenotype in QUESTIONNAIRE["phenotypes"].values():
        for q in phenotype["questions"]:
            for key, expr in q.get("computed", {}).items():
                result = _safe_eval(expr, enriched)
                if result is not None:
                    enriched[key] = result
    return enriched


# ---------------------------------------------------------------------------
# Step 2 - scoring engine
# ---------------------------------------------------------------------------
def score_phenotypes(answers: Dict[str, Any]) -> Dict[str, Any]:
    values = compute_derived_fields(answers)
    results = {}
    for pid, phenotype in QUESTIONNAIRE["phenotypes"].items():
        ticks = []
        biochemical_positive = 0
        for q in phenotype["questions"]:
            positive = evaluate_rule(q["rule"], values)
            if positive and q["category"] == "Biochemical Evaluation":
                biochemical_positive += 1
            ticks.append({"id": q["id"], "label": q["label"], "category": q["category"],
                          "reference": q["reference"], "positive": positive})
        total = len(phenotype["questions"])
        positive_count = sum(1 for t in ticks if t["positive"])
        results[pid] = {
            "label": phenotype["label"],
            "positive_count": positive_count,
            "total_questions": total,
            "percentage": round(positive_count / total * 100, 1),
            "severity": severity_label(positive_count),
            "biochemical_positive_count": biochemical_positive,
            "biomarker_support_flag": biochemical_positive >= 3,
            "ticks": ticks,
        }

    mito = QUESTIONNAIRE["mitochondrial_axis"]
    mito_ticks = [{"id": q["id"], "label": q["label"], "reference": q["reference"],
                   "positive": evaluate_rule(q["rule"], values)} for q in mito["questions"]]
    mito_positive = sum(1 for t in mito_ticks if t["positive"])
    results["mitochondrial"] = {
        "label": mito["label"],
        "positive_count": mito_positive,
        "total_questions": len(mito["questions"]),
        "percentage": round(mito_positive / len(mito["questions"]) * 100, 1),
        "ticks": mito_ticks,
        "flag": mito_positive >= 2,
    }
    return results, values


def severity_label(positive_count: int) -> str:
    for band in QUESTIONNAIRE["severity_bands"]["bands"]:
        if band["min"] <= positive_count <= band["max"]:
            return band["label"]
    return "Severe Driver / Primary Target"


# ---------------------------------------------------------------------------
# Step 3 - biomarker override rules
# ---------------------------------------------------------------------------
def evaluate_overrides(values: Dict[str, Any]) -> List[Dict[str, Any]]:
    triggered = []
    for override in QUESTIONNAIRE["overrides"]:
        if evaluate_rule(override["rule"], values):
            triggered.append({"id": override["id"], "phenotype": override["phenotype"],
                              "label": override["label"], "reference": override["reference"]})
    return triggered


# ---------------------------------------------------------------------------
# Step 4 - classification logic
# ---------------------------------------------------------------------------
def classify(scores: Dict[str, Any], overrides_triggered: List[Dict[str, Any]]) -> Dict[str, Any]:
    percentages = {pid: scores[pid]["percentage"] for pid in PHENOTYPE_IDS}
    max_pct = max(percentages.values())
    tally_leaders = [pid for pid, pct in percentages.items() if pct == max_pct]

    override_phenotypes = sorted(set(o["phenotype"] for o in overrides_triggered))
    biomarker_support = [pid for pid in PHENOTYPE_IDS if scores[pid]["biomarker_support_flag"]]

    if len(override_phenotypes) >= 2:
        classification = "Mixed/Combination Phenotype"
        primary = None
        involved = override_phenotypes
        reason = ("Multiple biomarker overrides triggered simultaneously (" +
                  ", ".join(o["label"] for o in overrides_triggered) +
                  ") -- Section 3 Key Biomarker Priority Overrules.")
    elif len(override_phenotypes) == 1:
        forced = override_phenotypes[0]
        classification = "Primary Phenotype (Biomarker Override)"
        primary = forced
        involved = [forced]
        matching = [o for o in overrides_triggered if o["phenotype"] == forced][0]
        reason = f"Biomarker override triggered: {matching['label']} -- {matching['reference']}."
        if forced not in tally_leaders:
            reason += f" This overrules the tally leader(s) ({', '.join(tally_leaders)})."
    elif max_pct == 0:
        classification = "Inconclusive / Minimal Findings"
        primary = None
        involved = []
        reason = "No positive ticks recorded in any of the four phenotype categories."
    elif len(tally_leaders) == 1:
        classification = "Primary Phenotype"
        primary = tally_leaders[0]
        involved = [primary]
        reason = f"Highest tick percentage ({max_pct}%) with no conflicting biomarker override."
    else:
        classification = "Mixed/Combination Phenotype"
        primary = None
        involved = tally_leaders
        reason = f"Tied highest tick percentage ({max_pct}%) between {', '.join(tally_leaders)} -- Section 3 Mixed/Combination rule."

    return {
        "classification": classification,
        "primary_phenotype": primary,
        "phenotypes_involved": involved,
        "reason": reason,
        "tally_leaders": tally_leaders,
        "overrides_triggered": overrides_triggered,
        "biomarker_supported_phenotypes": biomarker_support,
    }


def check_rotterdam_and_exclusions(answers: Dict[str, Any]) -> Dict[str, Any]:
    if answers.get("exclusion_other_disorders") == "yes":
        return {"status": "excluded", "reason": "Excluded due to other suspected thyroid or pituitary disorders."}
    
    rotterdam_count = 0
    if answers.get("irregular_cycle") == "yes":
        rotterdam_count += 1
    
    total_test = answers.get("total_testosterone")
    has_high_test = answers.get("high_testosterone_symptoms") == "yes" or (total_test is not None and float(total_test) > 45)
    if has_high_test:
        rotterdam_count += 1
        
    if answers.get("polycystic_ovaries_usg") == "yes":
        rotterdam_count += 1
        
    if rotterdam_count < 2:
        return {"status": "not_pcos", "reason": "Does not meet Rotterdam Criteria (requires 2 of 3: irregular cycles, high testosterone, polycystic ovaries)."}
        
    return {"status": "pcos", "reason": "Meets Rotterdam Criteria for PCOS diagnosis."}

def run_pipeline(answers: Dict[str, Any]) -> Dict[str, Any]:
    scores, enriched_values = score_phenotypes(answers)
    
    # Calculate bmi_calculated explicitly since safe_eval only handles simple binops
    weight = enriched_values.get("weight_kg")
    height = enriched_values.get("height_cm")
    if weight is not None and height is not None:
        try:
            enriched_values["bmi_calculated"] = float(weight) / (float(height) / 100) ** 2
        except ZeroDivisionError:
            pass

    rotterdam_result = check_rotterdam_and_exclusions(enriched_values)
    if rotterdam_result["status"] != "pcos":
        classification = {
            "classification": "Excluded" if rotterdam_result["status"] == "excluded" else "Inconclusive — Does not meet Rotterdam Criteria",
            "primary_phenotype": None,
            "phenotypes_involved": [],
            "reason": rotterdam_result["reason"],
            "tally_leaders": [],
            "overrides_triggered": [],
            "biomarker_supported_phenotypes": []
        }
        return {"scores": scores, "classification": classification, "computed_values": enriched_values}
        
    overrides_triggered = evaluate_overrides(enriched_values)
    classification = classify(scores, overrides_triggered)
    return {"scores": scores, "classification": classification, "computed_values": enriched_values}
